### PR TITLE
feat: initial implementation

### PR DESCRIPTION
**Pull request by jdiogopeixoto**
First implementation of `react-animate-text`.

### CODE REVIEW

File: README.md
Comment (fsdiogo) [inline README.md]: Just a suggestion: 

```suggestion
Many projects share the necessity of having animated text, so creating a component for that every time is a waste. This component offers a flexible solution to animate text that you can easily integrate and use in your project.
```

File: README.md
Comment (fsdiogo) [inline README.md]: We should take advantage of the `SplitText` component and add the ability to split the text by a [separator](https://github.com/moxystudio/react-split-text#separator) of our choosing.

It should be as simple as adding a `separator` prop and pass it the the `SplitText` component.

Comment (jdiogopeixoto) [reply README.md]: That is something that I thought about but ended up not adding. I will add it now!

File: README.md
Comment (fsdiogo) [inline README.md]: This default value should be `0` as it's the initial animation, should leave this as vanilla as possible.

File: README.md
Comment (fsdiogo) [inline README.md]: This should be `0.3` as it's our standard animation duration.

File: src/AnimateText.js
Comment (fsdiogo) [inline src/AnimateText.js]: Change the `initialDelay` and `lineDelay` to match the changes in the README:

```js
...
initialDelay: 0,
lineDelay: 0.3,
...
```

**Review by fsdiogo — APPROVED** [review]
Just change the default values, other than that LGMT 👍

File: package.json
Comment (acostalima) [inline package.json]:
```suggestion
  "description": "A react component that animates text per word and/or per line.",
```

File: src/AnimateText.js
Comment (acostalima) [inline src/AnimateText.js:27]: Can't we use `React.Children.forEach`?

File: src/AnimateText.js
Comment (acostalima) [inline src/AnimateText.js]: I wonder why we need these two props. Can't we just assume, by default, if `shouldShow` is `false` the text is displayed only when it's within the viewport? When `shouldShow` becomes `true`, the `IntersectionObserver` would become deactivated. 🤔

Comment (acostalima) [inline src/AnimateText.js]: Do we have any guideline on how these kind of props should be named, e.g. `show` vs `shouldShow`?

**Review by acostalima — CHANGES_REQUESTED** [review]

File: src/AnimateText.js
Comment (jdiogopeixoto) [reply src/AnimateText.js:27]: `React.Children` doesn't work here because we want to target the `containerRef.current.children` that already has the text split.

Comment (acostalima) [reply src/AnimateText.js:27]: 👍

File: src/AnimateText.js
Comment (jdiogopeixoto) [reply src/AnimateText.js]: I wouldn't do it because sometimes you might need to control the text with the `shouldShow` prop, but you still want the `IntersectionObserver` active so the animation only happens when the text is within the viewport. Does that make sense?

Comment (jdiogopeixoto) [reply src/AnimateText.js]: Since I extracted this from a project I just went with the same. Probably `isShowing` would be better here? What do you think?

Comment (acostalima) [reply src/AnimateText.js]: My understanding of how the component works is as follows:
- if `withObserver` is `true`, `shouldShow` does nothing be it `true` or `false`. The start of the animation is controlled by/delegated to the intersection observer.
- if `withObserver` is `false`, the intersection observer is disabled and the developer has to control `shouldShow`. When `shouldShow` changes to `true`, the text begins its animation
- in either cases above, the text is always rendered but the animation only kicks in when triggered by the intersection observer or by `shouldShow`.

Is the above correct?

Comment (acostalima) [reply src/AnimateText.js]: I prefer just `show`, similarly to the attributes of DOM elements, e.g., `disabled`, `checked`,  `hidden`, rather than `isDisabled`, etc.. However, more importantly than my personal preference is consistency with other libraries we've developed and agree on which path we should take. Shall we discuss this with the team?

Comment (jdiogopeixoto) [reply src/AnimateText.js]: You are correct, my bad! 

I guess one way to make it work with only one prop would be to check if the `shouldShow` is defined or not, and not have a default. When it is defined (no matter if `true` or `false`) we don't use the intersection observer, and let the developer control it. 

In the end that is what is happening now, the `shouldShow` prop is irrelevant unless the `withObserver` is set to `false`. Thoughts?

Comment (jdiogopeixoto) [reply src/AnimateText.js]: I am fine with `show` but you're right, I think that it is something that we can discuss with the team to keep the consistency amongst all the libraries.

Comment (fsdiogo) [reply src/AnimateText.js]: @acostalima I get your point of the DOM attributes and it's a good one, but in React I'm not a fan of using `show` or `disabled` as the type of value it holds is not explicit, it can either be a boolean or an object as far as I know. With `shouldShow`, `isDisabled`, etc. it's more explicit that it holds a boolean.

Comment (acostalima) [reply src/AnimateText.js]: ~🤔 From my understanding, all `shouldShow` does is activate the animation, right? The text is always rendered regardless of `shouldShow`'s value, even if we use the intersection observer. So, when `shouldShow` is `false` or `undefined` the behavior is exactly the same...
If `shouldShow` is irrelevant as you said when `withObserver` is `true`, why not just have one prop to determine whether the animation is controlled by the developer or automatically by the intersection observer? It's illogical for both behaviors to coexist anyway, correct?~

Your approach should work! 👌 
If `withObserver` is `false`, it is supposed to be possible to restart the animation by changing `shouldShow` as follows: `true` -> `false` -> `true`?

Comment (acostalima) [reply src/AnimateText.js]: @fsdiogo Well, in this particular context, being `show` both a verb and a name, I'd never say `show` is an object. A function? More likely. My point here is that I've seen developers doing your way and other's mine. We should put our personal preferences aside and reach a consensus as a team. I don't want to block the PR because of this, but please do start a discussion on Slack. When we have a decision, the repository's maintainer (@jdiogopeixoto) should follow up by opening an issue or PR to address the issue, if applicable. 💪 
Sounds good?

Comment (ruipneves) [reply src/AnimateText.js]: I personally do not like `shouldShow` prop name, but it is functional and works. Personally I prefer something like `isVisible` but to be honest I don't really see the problem here. As long as the prop name states what it does or represents it in a clear way I'm fine with either name. I think we don't need to be that picky.

Comment (zebateira) [reply src/AnimateText.js]: I understand what @fsdiogo is saying that a good rule of thumb for naming booleans is to prefix them with `is*` or `should*`, but in this case we can argue that `show` is a good exception to this rule considering the context of this package which is to show or hide the text by animating it. If the package had more complex options the naming of the props might have to follow the rule to avoid ambiguous names. But in this case I think it's fine to just use `show` - you'd first assume it's a boolean considering you are using this package to show and hide text.

I don't think we need to apply the same naming rule to every single package, but rather find a common ground on what makes sense to be the same or different.

Let me know what you think.

Comment (jdiogopeixoto) [reply src/AnimateText.js]: Yes, but we would remove the `withObserver` prop. So the only way to see if the text is being controlled or not would be by checking if the `shouldShow` prop is defined or not - we need it on [this line](https://github.com/moxystudio/react-animate-text/pull/1/files#diff-77a05e19585bf2c1a6d7c5a9529d0a32R68).

Comment (acostalima) [reply src/AnimateText.js]: @jdiogopeixoto 👌 
I think it's better than having a prop which does nothing in one specific use case. I'd probably rename the prop to `show` then. It's not suitable to prefix it with `is` either assuming the prefix should only be used for props that only accept either `true` or `false`.

Comment (acostalima) [reply src/AnimateText.js]: @zebateira sounds reasonable!

File: README.md
Comment (acostalima) [inline README.md:86]:
```suggestion
Type: `boolean` | Required: `false` | Default: `undefined`
```

**Review by acostalima — APPROVED** [review]

**Comment by acostalima** [comment]
@jdiogopeixoto, do you want me to merge this?

**Comment by jdiogopeixoto** [comment]
@acostalima Sure, it is good to go 🚀

**Pull request merged**
